### PR TITLE
Fix #955 -- add description of deployment procedure

### LINKED ISSUE #955: Deployment on production

### PR DESCRIPTION
**Pull request by chrismedrela**
Fixes #955.

### CODE REVIEW

**Comment by chrismedrela** [comment]
@pbanaszkiewicz please review.

File: docs/deploy.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
+# Procedure for releasing new AMY version
+
+1. Close milestone using [GitHub UI](https://github.com/swcarpentry/amy/milestones).
+
+1. Follow procedure for deployment on production.
+
+1. Write down release notes and publish it on [Piotr's blog](http://piotr.banaszkiewicz.org/). Release notes should have two sections: new features and bugfixes, each one enumerating changes and mentioning their authors. Base your work on the list of all closed pull request for given milestone, available on GitHub.
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: 1. "all closed pull request**s**"
2. Can you split long lines? It will be easier to comment on single lines.

File: docs/deploy.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            $ git pull upstream develop
+            $ git push origin develop
+            
+            $ git checkout master
+            $ git pull upstream master
+            $ git push origin master
+        
+        Pushes to your `origin` repository are optional.
+        
+    1.  Update `master` branch:
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: Merge `develop` into `master` branch.

File: docs/deploy.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    
+    1.  Just to be safe, run tests.
+    
+            $ make test
+    
+    1.  Tag a release. Tag message should be `AMY release v1.X.Y`. 
+        
+            $ git tag -a v1.7.1 -s
+            
+        Omit `-s` flag if you cannot create signed tags.
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: Can you please add a link to some help regarding signed tags here? thanks

File: docs/deploy.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            $ git add workshops/__init__.py bower.json
+            $ git commit -m "Bumping version to v1.X.Y"
+    
+    1.  Just to be safe, run tests.
+    
+            $ make test
+    
+    1.  Tag a release. Tag message should be `AMY release v1.X.Y`. 
+        
+            $ git tag -a v1.7.1 -s
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: In [line 52](https://github.com/swcarpentry/amy/pull/1077/files#diff-ef860daacd753531645fa756fbe5b836R52) you have `v1.X.Y` but here you use explicit values for X and Y: `v1.7.1`. Maybe it's a good idea to keep X and Y everywhere or the same example `v1.7.1` everywhere?

File: docs/deploy.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    
+    1.  Push `master` and the new tag everywhere:
+    
+            $ git push origin master --tags
+            $ git push upstream master --tags
+    
+1.  Bump version on `develop`:
+
+        $ git checkout develop
+        $ vim bower.json  # change version to the next one after 1.X.Y (i.e. 1.X+1.Y)
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: That would be `1.X+1.0`.

Additional note: we won't have major version ==1 forever.

File: docs/deploy.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            ssh amy@amy.software-carpentry.org
+
+        All the following commands will be executed on production, not your local machine.
+        
+    1.  Execute:
+        
+            source ~/amy_site/venv/bin/activate
+            cd amy_site/amy
+            git fetch
+            git checkout tags/v1.7.1  # Python files updated
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: "Python files updated, but not the served files - server keeps them in memory" or something similar

File: docs/deploy.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            
+            $ cp db.sqlite3 db.sqlite3.bak
+            
+    1.  Update dependencies:
+            
+            $ make upgrade
+            
+    1.  Migrate database online:
+    
+            $ ./manage.py migrate
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: What about `AMY_DB_NAME`?

File: docs/deploy.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            
+    1.  Regenerates version in the footer:
```
Comment (pbanaszkiewicz) [inline docs/deploy.md]: "version number".

This is easier than writing `make workshops/git_version.py` plus it runs checks.

**Comment by chrismedrela** [comment]
@pbanaszkiewicz I've made changes you requested. I don't know if the way I added `AMY_DB_FILENAME` is correct, but I cannot verify it and I don't want to experiment on production now.

> I think it's worth to have an additional step that tries to apply migrations on a production database copy.

I don't think so - I think we should temporarily stop server for upgrade. When we do that, it doesn't matter if you apply migrations to `db.sqlite3` or its copy, because the server is off and none of these dbs can change during migration.

**Comment by pbanaszkiewicz** [comment]
@chrismedrela:

> I don't know if the way I added `AMY_DB_FILENAME` is correct, but I cannot verify it and I don't want to experiment on production now.

It's not correct, because you're setting it in your own environment, not server's (or, not uWSGI's, to be precise). `AMY_DB_FILENAME`, along with other envvars, is set up in `amy_uwsgi.ini` config file.

>> I think it's worth to have an additional step that tries to apply migrations on a production database copy.

> I don't think so - I think we should temporarily stop server for upgrade. When we do that, it doesn't matter if you apply migrations to `db.sqlite3` or its copy, because the server is off and none of these dbs can change during migration.

I changed a few things on the production. First of all, I added a `/etc/init` (or Upstart) script for running uWSGI in Emperor mode. This way admin can start/stop uWSGI easily (`start uwsgi-emperor` and `stop uwsgi-emperor`). I also altered `nginx` configuration so that it will show maintenance page `/home/amy/amy_site/static_html/maintenance.html` when `uwsgi` is not responding.

There is one caveat, though. "Normal" user cannot issue `start/stop uwsgi-emperor` commands, only admin can. If anyone has a solution to this issue, I'll be grateful, because for now we cannot stop the server for upgrade.

While I agree with you that being able to stop uWSGI for AMY upgrade is nice, I didn't intend to use `AMY_DB_FILENAME` because the DB would change during migration; instead I was thinking about migrations failing. It happened a few times, mostly due to data migrations and not all cases covered. I don't think this would be the case if we could stop uWSGI for the upgrade.

Finally I noticed that we should be running `./manage.py` with the same environmental variables as we run the production; this is something we should think about…

Anyway, to sum up:

1. changes on the production to: start/stop uWSGI easily, and have a maintenance page show up when uWSGI is off
2. need to be able to start/stop uWSGI by normal user (`amy`), not only by root
3. need to read and set up envvars from `amy_uwsgi.ini` either when sourcing the virtualenv or when issuing any `./manage.py` command
4. also, the part about `AMY_DB_FILENAME` will have to change.

**Comment by chrismedrela** [comment]
> 3. need to read and set up envvars from amy_uwsgi.ini either when sourcing the virtualenv or when issuing any ./manage.py command

According to [this](https://github.com/unbit/uwsgi/issues/629#issuecomment-43303330), you can factor out envvars to `.env` file, and modify `amy_uwsgirc.ini`:
```
[uwsgi]
for-readline = ../.env
  env = %(_)
endfor = 
```

**Comment by chrismedrela** [comment]
@pbanaszkiewicz please review again.
